sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -5816,20 +5823,40 @@ pi_result piEnqueueEventsWaitWithBarrier(pi_queue Queue,
   std::scoped_lock<ur_shared_mutex> lock(Queue->Mutex);
 
   // Helper function for appending a barrier to a command list.
-  auto insertBarrierIntoCmdList =
-      [&Queue](pi_command_list_ptr_t CmdList,
-               const _pi_ze_event_list_t &EventWaitList, pi_event &Event,
-               bool IsInternal) {
-        if (auto Res = createEventAndAssociateQueue(
-                Queue, &Event, PI_COMMAND_TYPE_USER, CmdList, IsInternal))
-          return Res;
+  auto insertBarrierIntoCmdList = [&Queue](
+                                      pi_command_list_ptr_t CmdList,
+                                      const _pi_ze_event_list_t &EventWaitList,
+                                      pi_event &Event, bool IsInternal) {
+    if (auto Res = createEventAndAssociateQueue(
+            Queue, &Event, PI_COMMAND_TYPE_USER, CmdList, IsInternal))
+      return Res;
 
-        Event->WaitList = EventWaitList;
-        ZE_CALL(zeCommandListAppendBarrier,
-                (CmdList->first, Event->ZeEvent, EventWaitList.Length,
-                 EventWaitList.ZeEventList));
-        return PI_SUCCESS;
-      };
+    Event->WaitList = EventWaitList;
+
+    // For in-order queue we don't need a real barrier, just add a "barrier"
+    // event signal because it is already guaranteed that previous commands
+    // are completed when the signal is started.
+    // TODO: this and other special handling of in-order queues to be
+    // updated when/if Level Zero adds native support for in-order queues.
+    //
+    if (Queue->isInOrderQueue() && InOrderBarrierBySignal) {
+      ZE_CALL(zeCommandListAppendSignalEvent, (CmdList->first, Event->ZeEvent));
+    } else {
```

Comment:
Sorry, but as far as I understand this will not work for all cases.

a. For out-of-order queue we have two scenarios:
1. When EventWaitList provided to piEnqueueEventsWaitWithBarrier is not empty (NumEventsInWaitList != 0). In this case we need to insert a single barrier which is waiting for events from EventWaitList. This is the corresponding call: https://github.com/intel/llvm/blob/sycl/sycl/plugins/level_zero/pi_level_zero.cpp#L5867-L5869

NB. When event list is provided - barrier has to wait only for provided wait list and it's not a queue-wide barrier.


3. When EventWaitList provided to piEnqueueEventsWaitWithBarrier is empty (NumEventsInWaitList == 0). 
In this case we need to insert queue-wide barrier. I.e. for out-of-order queue we insert a barrier to multiple command lists corresponding to each level zero queue. And we also insert a convergence barrier.

b. Currently for in-order queues we always take this branch:
https://github.com/intel/llvm/blob/sycl/sycl/plugins/level_zero/pi_level_zero.cpp#L5851
I.e. it is some kind of optimization because for in-order queue we always insert 1 barrier only (in both scenarios described above).
 
If NumEventsInWaitList == 0 then we insert a barrier which wairt for the last command event,
If we NumEventsInWaitList != 0 then we insert a barrier which waits for EventWaitList and the last command event.

Problem is that with this new change we will just signal some event.
1. if NumEventsInWaitList != 0 then we will not wait for any event from EventWaitList  (they may be even from different queues).
2. If NumEventsInWaitList == 0 then we will not wait for the last command event. Because inserted zeCommandListAppendSignalEvent doesn't depend on the last command event. So, if  someone insert a barrier and waits for an output event - there is not guarantee that the last event or any other event in the in-order queue is waited.

Reply:
@againull : Good catch! What I really wanted to do is to wait last command and then signal the event that would represent the barrier. But I am indeed not "waiting" with the current changes. And I see no L0 API to wait + signal. Perhaps, then I could just use L0 event from the last command and just use that, wrapped into a PI event. And is there are no last command then just do what I do now, signal a new event because there is nothing to wait for. Would that work?

@jandres742 : what do you think?

Reply:
Yes, this sounds good for the NumEventsInWaitList == 0 case. 
a. If there is LastCommandEvent and it is not internal (if it is internal we can't do that I guess) then we probably can just return it as output event  and don't create barrier or signal any new event.
b. If  LastCommandEvent = nullptr, then as you have said we signal and return.
c. If NumEventsInWaitList != 0 we either need to insert a barrier waiting for all events in the waitlist  - because they may be from different queue. Or we can check the queue for all events there and somehow optimize using a. and .b approaches.

Reply:
> need to insert a barrier waiting for all events in the waitlist - because they may be from different queue. 

They can be from different L0 queues, but still from the same SYCL queue (a single SYCL queue may span across multiple queues). And then I think we should not need to do anything else but piggy-backing on the `LastCommandEvent`. It is then guaranteed by the in-order dependencies that all previous commands (in all L0 queues used by this SYCL queue) are completed before `LastCommandEvent`. Am I missing something?

Reply:
For some reasons I had an impression that EventList provided to piEnqueueEventsWaitWithBarrier can contain events from different SYCL/PI queues.

Reply:
I think this API corresponds to https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_enqueue_barrier.asciidoc

void ext_oneapi_barrier( const std::vector<event> &waitList );

So my understanding was that any sycl event from any queue can be provided there.

Reply:
> So my understanding was that any sycl event from any queue can be provided there.

OK, I think you are right. Re-working.

Reply:
@againull , @jandres742 : please review